src/util/readpool/mod.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod config;
+mod context;
+mod priority;
+
+use std::time;
+use futures::Future;
+use futures_cpupool as cpupool;
+
+use util;
+use util::futurepool;
+
+pub use self::config::Config;
+pub use self::context::Context;
+pub use self::priority::Priority;
+
+const TICK_INTERVAL_SEC: u64 = 1;
+
+pub struct ReadPool {
+    pool_high: futurepool::FuturePool<Context>,
+    pool_normal: futurepool::FuturePool<Context>,
+    pool_low: futurepool::FuturePool<Context>,
+}
+
+impl util::AssertSend for ReadPool {}
```

Comment:
why use AssertSend here, why not use Send directly?

Reply:
When `impl Send`, the compiler won't check whether it is indeed Send. Instead it will trust the developer and mark the struct as Send anyway (so that `impl Send` is unsafe). I use `AssertSend` here to ensure that `ReadPool` is Send. In fact I borrowed this pattern from [futures-cpupool library](https://github.com/alexcrichton/futures-rs/blob/master/futures-cpupool/src/lib.rs#L95).